augur/filter.py

```diff
@@ -71,6 +69,30 @@ def resulting_sequences(self):
             filtered_sequences = filter_obj.filter(filtered_sequences, all_sequences)
 
         return filtered_sequences
+    """
+
+    def filter_sequences(self):
+        self.resulting_sequences = [
+            sample
+            for sample in self.sequences
+            if self.apply_filter_chain(sample) == "include"
+        ]
+
+    # TODO new class FilterChain? also rename all of these `sequence` arguments to `sample`
+    def apply_filter_chain(self, sample):
+        result = "include"
+
+        for filter_obj in self.filter_chain:
+            if result == filter_obj.FILTER_OPERATION:
+                # If the result already matches the current filter's operation, skip it. This saves us from
+                # doing expensive calculations to decide to exclude a sample that has already been excluded
+                # by an earlier filter.
+                continue
+
+            if filter_obj.is_affected(sample):
+                result = filter_obj.FILTER_OPERATION
+
+        return result
```

Comment:
I really don't like the string comparison here.

"include" vs "exclude" is boolean: include == True, exclude == False

Reply:
I think I prefer your `Enum` suggestion from yesterday (though see my future comment below about just having `filters.EXCLUDE_FILTERS` and `filters.INCLUDE_FILTERS`), but I just want to point out:

As I understand it, strings in Python are automatically interned. Every occurrence of a specific string literal is a pointer to a a single, immutable object representing that sequence of characters. So `result = filter_obj.FILTER_OPERATION` is actually not a string comparison like C's `strcmp()`, but a pointer comparison to test whether the two references point to the same interned object. It's conceptually equivalent to using symbols in Ruby.

I'm not 100% sure on that; Python isn't my primary language. See also https://medium.com/@bdov_/https-medium-com-bdov-python-objects-part-iii-string-interning-625d3c7319de.

The chief advantage of using an Enum, in my opinion, is that the compiler will detect typos—`return "exclud"` might be bothersome to track down.

Reply:
But why use Include and Exclude at all, regardless of enum vs string? They're the same operation.

Reply:
Right. See https://github.com/nextstrain/augur/pull/494#discussion_r401845763.